File: backend/services/questions.py
import asyncio
import json
import logging
import random
from pathlib import Path

from services.cache import TTLPool
from services.groq_client import generate_questions
from services.images import fetch_image_url
from services.safety import filter_safe
from services.wikipedia_fallback import generate_from_wikipedia

logger = logging.getLogger("kidgk.questions")

DATA_DIR = Path(__file__).resolve().parent.parent / "data"

with open(DATA_DIR / "categories.json", encoding="utf-8") as f:
    CATEGORIES = json.load(f)

with open(DATA_DIR / "fallback_questions.json", encoding="utf-8") as f:
    FALLBACK_BANK = json.load(f)

with open(DATA_DIR / "fallback_questions_expanded.json", encoding="utf-8") as f:
    FALLBACK_BANK_EXPANDED = json.load(f)


def _offline_pool(category_id: str) -> list[dict]:
    expanded = FALLBACK_BANK_EXPANDED.get(category_id) or []
    return expanded if expanded else FALLBACK_BANK.get(category_id, [])

CATEGORY_IDS = {c["id"] for c in CATEGORIES}

# Generated question sets are cached per category+grade and reused across
# sessions/rounds — cuts Groq calls roughly in line with proposal §5's
# ~80% reduction target while staying well within free-tier rate limits.
POOL_SIZE = 20
CACHE_TTL_SECONDS = 6 * 60 * 60  # 6 hours

_pool_cache: TTLPool[list[dict]] = TTLPool(CACHE_TTL_SECONDS)


def fallback_round(category_id: str, count: int) -> list[dict]:
    pool = _offline_pool(category_id)
    picked = random.sample(pool, min(count, len(pool)))
    return picked


async def _attach_images(questions: list[dict]) -> list[dict]:
    urls = await asyncio.gather(
        *(fetch_image_url(q.get("image_keyword")) for q in questions)
    )
    for q, url in zip(questions, urls):
        q["image_url"] = url
    return questions


async def _get_pool(category_id: str, category_name: str, grade: int) -> tuple[list[dict], str]:
    """Returns a cached (or freshly generated) pool of questions plus its source.
    Three tiers, in order: Groq (fresh, cached) -> Wikipedia-derived facts
    (cached, so a Groq outage/rate-limit doesn't re-hit Wikipedia every
    request) -> static offline bank (never cached, always retries the live
    tiers next time)."""
    key = f"{category_id}:{grade}"
    cached = _pool_cache.get(key)
    if cached is not None:
        return cached, "cache"

    async with _pool_cache.lock_for(key):
        cached = _pool_cache.get(key)  # re-check after acquiring the lock
        if cached is not None:
            return cached, "cache"

        try:
            raw = await generate_questions(category_name, grade, POOL_SIZE)
            filtered = filter_safe(raw)
            if len(filtered) < min(5, POOL_SIZE):
                raise RuntimeError(
                    f"Only {len(filtered)}/{POOL_SIZE} Groq questions passed validation/safety checks"
                )
            _pool_cache.set(key, filtered)
            return filtered, "groq"
        except Exception as exc:
            logger.warning("Groq unavailable for %s (%s) - trying Wikipedia", category_id, exc)

        try:
            wiki = await generate_from_wikipedia(category_id, POOL_SIZE)
            filtered = filter_safe(wiki)
            if len(filtered) < min(5, POOL_SIZE):
                raise RuntimeError(
                    f"Only {len(filtered)}/{POOL_SIZE} Wikipedia questions passed safety checks"
                )
            _pool_cache.set(key, filtered)
            return filtered, "wikipedia"
        except Exception as exc:
            logger.warning("Wikipedia fallback failed for %s (%s) - using offline bank", category_id, exc)
            return _offline_pool(category_id), "fallback"


async def get_round(category_id: str, grade: int, count: int = 5) -> dict:
    if category_id not in CATEGORY_IDS:
        raise ValueError(f"Unknown category: {category_id}")

    category_name = next(c["name"] for c in CATEGORIES if c["id"] == category_id)

    pool, source = await _get_pool(category_id, category_name, grade)
    questions = random.sample(pool, min(count, len(pool)))
    questions = await _attach_images(questions)

    return {"category": category_id, "source": source, "questions": questions}

File: frontend/src/lib/questions.ts
import type { Question, Round, RoundSource } from "./types";
import { CATEGORIES } from "./categories";
import { FALLBACK_BANK } from "./fallbackQuestions";
import { generateQuestions } from "./groq";
import { generateFromWikipedia } from "./wikipediaFallback";
import { filterSafe } from "./safety";
import { fetchImageUrl } from "./images";

// Generated question pools are cached per category+grade for the lifetime
// of the page (mirrors backend/services/cache.py's TTL pool, minus the TTL
// since a browser tab doesn't stay open for hours the way a server does).
const POOL_SIZE = 20;
const poolCache = new Map<string, Question[]>();
const inFlight = new Map<string, Promise<{ pool: Question[]; source: RoundSource }>>();

function sample<T>(arr: T[], n: number): T[] {
  const a = [...arr];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a.slice(0, n);
}

// The expanded bank (~800 questions, ~600KB) is only needed on the rare
// path where both Groq and Wikipedia have failed, so it's dynamically
// imported as its own chunk instead of bloating the main bundle every
// visitor downloads regardless of whether they ever hit this tier.
async function offlinePool(categoryId: string): Promise<Question[]> {
  try {
    const { FALLBACK_BANK_EXPANDED } = await import("./fallbackQuestionsExpanded");
    const expanded = FALLBACK_BANK_EXPANDED[categoryId];
    if (expanded && expanded.length > 0) return expanded;
  } catch {
    // fall through to the small bundled bank if the chunk fails to load
  }
  return FALLBACK_BANK[categoryId] ?? [];
}

async function fallbackRound(categoryId: string, count: number): Promise<Question[]> {
  const pool = await offlinePool(categoryId);
  return sample(pool, Math.min(count, pool.length));
}

async function attachImages(questions: Question[]): Promise<Question[]> {
  const urls = await Promise.all(questions.map((q) => fetchImageUrl(q.image_keyword)));
  return questions.map((q, i) => ({ ...q, image_url: urls[i] }));
}

// Primary tier: the pre-generated static bank (tools/build_groq_bank.py),
// keyed by category:grade and committed to the repo. Instant, high quality,
// and immune to Groq's shared daily rate limit - each round is a fresh
// random sample from the ~30 questions per cell. Lazily imported as its
// own chunk to keep the main bundle small.
async function staticBankPool(categoryId: string, grade: number): Promise<Question[] | null> {
  try {
    const { GROQ_BANK } = await import("./groqBank");
    const pool = GROQ_BANK[`${categoryId}:${grade}`];
    return pool && pool.length >= POOL_SIZE ? pool : null;
  } catch {
    return null;
  }
}

async function getPool(
  categoryId: string,
  categoryName: string,
  grade: number
): Promise<{ pool: Question[]; source: RoundSource }> {
  // The static bank is not cached in poolCache on purpose: sampling the
  // full 30-question cell fresh each round is what gives round-to-round
  // variety ("put things randomly").
  const bankPool = await staticBankPool(categoryId, grade);
  if (bankPool) return { pool: bankPool, source: "bank" };

  const key = `${categoryId}:${grade}`;
  const cached = poolCache.get(key);
  if (cached) return { pool: cached, source: "cache" };

  const existing = inFlight.get(key);
  if (existing) return existing;

  const promise = (async () => {
    try {
      const raw = await generateQuestions(categoryName, grade, POOL_SIZE);
      const filtered = filterSafe(raw);
      if (filtered.length < Math.min(5, POOL_SIZE)) {
        throw new Error(
          `Only ${filtered.length}/${POOL_SIZE} Groq questions passed validation/safety checks`
        );
      }
      poolCache.set(key, filtered);
      return { pool: filtered, source: "groq" as RoundSource };
    } catch {
      // fall through to Wikipedia
    }

    try {
      const wiki = await generateFromWikipedia(categoryId, POOL_SIZE);
      const filtered = filterSafe(wiki);
      if (filtered.length < Math.min(5, POOL_SIZE)) {
        throw new Error(
          `Only ${filtered.length}/${POOL_SIZE} Wikipedia questions passed safety checks`
        );
      }
      poolCache.set(key, filtered);
      return { pool: filtered, source: "wikipedia" as RoundSource };
    } catch {
      return { pool: await offlinePool(categoryId), source: "fallback" as RoundSource };
    }
  })();

  inFlight.set(key, promise);
  try {
    return await promise;
  } finally {
    inFlight.delete(key);
  }
}

function normalize(text: string): string {
  return text.trim().toLowerCase();
}

// The static bank shares identical question pools across some adjacent
// grades within the same difficulty band (e.g. grade 4 and 5 use the same
// 30 questions per category), so a mid-round difficulty adjustment can land
// on a pool that overlaps with what was already asked. excludeQuestions lets
// the caller filter those out before sampling, so a round never repeats a
// question the student has already seen.
export async function getRound(
  categoryId: string,
  grade: number,
  count = 5,
  excludeQuestions: string[] = []
): Promise<Round> {
  const category = CATEGORIES.find((c) => c.id === categoryId);
  if (!category) throw new Error(`Unknown category: ${categoryId}`);

  const { pool, source } = await getPool(categoryId, category.name, grade);
  const exclude = new Set(excludeQuestions.map(normalize));
  const available = exclude.size > 0 ? pool.filter((q) => !exclude.has(normalize(q.question))) : pool;
  const usable = available.length >= Math.min(count, pool.length) ? available : pool;
  const questions = await attachImages(sample(usable, Math.min(count, usable.length)));

  return { category: categoryId, source, questions };
}

// Exported for completeness/parity with the old backend's fallback_round;
// unused directly since getPool already falls through to FALLBACK_BANK.
export { fallbackRound };
